zio-http-benchmarks/src/main/scala/zhttp.benchmarks/ApiBenchmark.scala

```diff
@@ -0,0 +1,124 @@
+package zio.benchmarks
+
+import org.openjdk.jmh.annotations._
+import zio.http._
+import zio.http.model.Method
+import zio.json.{DeriveJsonCodec, EncoderOps, JsonCodec}
+import zio.schema.{DeriveSchema, Schema}
+import zio.{Scope => _, _}
+
+import java.util.concurrent.TimeUnit
+
+final case class ExampleData(userId: Int, postId: Int, query: String)
+
+object ExampleData {
+  implicit val schema: Schema[ExampleData] =
+    DeriveSchema.gen[ExampleData]
+
+  implicit val jsonCodec: JsonCodec[ExampleData] =
+    DeriveJsonCodec.gen[ExampleData]
+}
+
+@State(Scope.Thread)
+@BenchmarkMode(Array(Mode.Throughput))
+@OutputTimeUnit(TimeUnit.SECONDS)
+class ApiBenchmark {
+
+  private val REPEAT_N = 1_000
+
+  import zio.http.api._
+
+  // # Small Data Request
+
+  // API DSL
+  val usersPosts =
+    API
+      .get(In.literal("users") / In.int / In.literal("posts") / In.int)
+      .in(In.query("query"))
+      .out[ExampleData]
+
+  val handledUsersPosts =
+    usersPosts.handle { case (userId, postId, limit) =>
+      ZIO.succeed(ExampleData(userId, postId, limit))
+    }
+
+  val apiHttpApp = handledUsersPosts.toHttpApp
+
+  // Collect DSL
+  val collectHttpApp = Http.collectZIO[Request] { //
+    case req @ Method.GET -> !! / "users" / userId / "posts" / postId =>
+      val userIdInt = userId.toInt
+      val postIdInt = postId.toInt
+      val query     = req.url.queryParams.get("query").flatMap(_.headOption).get
+      ZIO.succeed(
+        Response.json(
+          ExampleData(userIdInt, postIdInt, query).toJson,
+        ),
+      )
+  }
+
+  val request: Request =
+    requestFromString("http://localhost:8080/users/1/posts/2?query=cool")
+
+  def unsafeRun[E, A](zio: ZIO[Any, E, A]): Unit = Unsafe.unsafe { implicit unsafe =>
+    Runtime.default.unsafe
+      .run(zio.unit)
+      .getOrThrowFiberFailure()
+  }
+
+  @Benchmark
+  def benchmarkSmallDataZioApi(): Unit =
+    unsafeRun {
+      apiHttpApp(request).repeatN(REPEAT_N)
+    }
+
+  @Benchmark
+  def benchmarkSmallDataZioCollect(): Unit =
+    unsafeRun {
+      collectHttpApp(request).repeatN(REPEAT_N)
+    }
+
+  // # Large Path
+
+  // API DSL
+  import In._
+
+  val largePathHttpApp = API
+    .get(In.literal("first") / int / "second" / int / "third" / int / "fourth" / int / "fifth" / int / "sixth" / int)
```

Comment:
This is a "deep path", which is great to benchmark!

In addition to the deep path, we should have a "broad path". There should be 1000 endpoints that arrange nicely into a tree: 10 wide by 10 deep. 

`O(n)` matchers will fair poorly in this case, but the prefix tree should fair quite well, all other things being equal.

The benchmark should hit a random API out of those 1000 endpoints just to be fair.

Reply:
We didn't get to 1000 endpoints yet, but on a semi-realistic set of broad routes, we're already starting to pull ahead:

```scala
[info] ApiBenchmark.benchmarkBroadTapirAkka        thrpt    4   78.794 ±  61.142  ops/s
[info] ApiBenchmark.benchmarkBroadTapirHttp4s      thrpt    4   86.456 ±  50.664  ops/s
[info] ApiBenchmark.benchmarkBroadZioApi           thrpt    4  667.506 ± 575.568  ops/s <-----
[info] ApiBenchmark.benchmarkBroadZioCollect       thrpt    4  119.175 ±  30.864  ops/s

[info] ApiBenchmark.benchmarkDeepPathTapirAkka     thrpt    4   39.212 ±  52.063  ops/s
[info] ApiBenchmark.benchmarkDeepPathTapirHttp4s   thrpt    4   42.098 ±   5.846  ops/s
[info] ApiBenchmark.benchmarkDeepPathZioApi        thrpt    4  433.606 ± 708.764  ops/s
[info] ApiBenchmark.benchmarkDeepPathZioCollect    thrpt    4  855.360 ± 342.678  ops/s

[info] ApiBenchmark.benchmarkSmallDataTapirAkka    thrpt    4   65.132 ±   5.211  ops/s
[info] ApiBenchmark.benchmarkSmallDataTapirHttp4s  thrpt    4   69.190 ±   3.894  ops/s
[info] ApiBenchmark.benchmarkSmallDataZioApi       thrpt    4  280.864 ±   5.790  ops/s
[info] ApiBenchmark.benchmarkSmallDataZioCollect   thrpt    4  681.826 ± 241.145  ops/s
```

Reply:
The `Collect` is using zio-http's `Http.collectZIO[Request]` with pattern matching. I think we should be able to match/beat that in the other benchmarks after further optimization.